README.md
An AppleII+ emulator in C++14
===================

###Goals:
-	Simple code, easy to follow and change, somewhat performant.
-	Modern C++ coding style followed as much as possible.

###Caveats:
-	Emulates only an Apple2+. Does not write to floppy disks.
-	CPU	Does not implement decimal mode (yet).

###3 Layers:

1 - Core : All the 6502 emulation. Doesn't not have any idea of how an apple2 is supposed to work
	- A clock
	- A cpu (which could probably be redesigned in a better way)
	- A bus and various chips (ram, rom, i/o)

2 - a2 : All the emulation of the following apple2+ components:
	* The keyboard
	* The video
	* The speaker
	* The floppy disk (read-only)
	* The i/o memory
	In addition, it also contains an "Apple2" class that can create an apple2
	=> This implements the hardware, but have no UI per see

3 - sdl : An sdl layer on top of the a2 component that brings the emulator to life. It contains
	- A keyboard input
	- A video renderer
	- A sound output
	- A main emulation loop
	=> Implement the look and display

Each layer is only dependant on the ones before.

All the layers are bound together with an internal command-line interface that let access the emulator while running. By default, pressing F1 will drop into that command line.

TODO:
-	gif demo in github
-	emscriptem demo in github
-	Fix command-line mess
-	OSX back-port ?
-	Windows port ?
-	iOS port ?
-	latch object ?
-	clean-up cpu_6502 ?
-	JIT ?


HISTORY.md
I started writing this code on Sunday May, 1st, 2016 at 1AM. I had always wanted to write an emulator, so I sat down and said "let's write a Z80 emulator".

A couple of minutes later, I asked myself "what software do you want to run on it", and decided to do a 6502, so I could test it on parts of the Apple 2 ROM (also, I own an Apple2+ in my bsement, even if I haven't try to boot it in the last 20 years).

I know about nothing on 6502, but thought it would be easier anyway. I think it was the right decision, as the amount of documentaiton saved my life more than once.

The few design decisions, at this point were to use C++, and to use templates to have the compiler generate the correct code for the instructions. I don't what the absolute best performance, but I do want it to be decent (so I decided to avoid have too many indirections around). Also, it seems like a good modern C++ template use.

Seeing that I had to execute the ROM code, I needed to have the apple2 memory layout (ie: to have the code loaded at the right place). I used an existing emulator to check that the ROM I found online was more or less correct, and downloaded two documentations:

* Programming the 65816 (Authors David Eyes and Ron Lichty)

and

* The Apple II Reference Manual (Apple Computer Inc.)

A - The initial code

The initial code was trivial: create a 64K block of memory, load the 0x3000 bytes of the ROM at address 0xd000 and implement reset on my 6502 CPU.

void reset()
{
	PC = memory.read_word( 0xfffc );
	S = 0xff;
}

While this may look trivial, finding the right information is sometimes a challenge, and the CPU reference manual never directly tells you "At RESET, the CPU puts $FF in the stack regiser, loads the memory word at $FFFC and starts from there".

In the code, registers are implemented via ivars of the exact same name as the 6502, and the CPU knows a memory object that implement read_byte, read_word, write_byte and write_word. byte is a typedef to unsigned char, and word is a public C++ class with a H and L member:

struct word
{
	byte h;
	byte l;

	word( int v ) : h( v>>8 ), l(v) {}
	word( byte h, byte l ) : h(h), l(l) {}

	operator int() const { return h*0x100+l; }
	word operator+( int v ) const { return ((int)(*this))+v; }
};

The code started its life as a loop that just said:

while (1)
{
	byte opcode = memory.read_byte( PC );
	PC = PC + 1;
	switch (opcode)
	{
		default:
			std::out << "Unimplemented opcode $" << to_hex2(opcode) << " at $" << to_hex4(PC-1) << "std::endl;
			assert( false );
			break;
	}
}

and the first output was:

Unimplemented opcode $D8 at $FA62.

B - Implementing opcodes, one after the other

From then on, going forward is extremely simple. Look at next unimplemented instruction in 6502 manual, implement it, run again.

For CLD, the implementation looked like:

set_flag( DECIMAL, 0 );
cycles += 2;
std::stdout << "STD" << std::endl;

DECIMAL is an integer that represent the position of the Decimal flag in the P status register:

	const int DECIMAL = 3;

	void set_flag( int f ) { P |= 1<<f; }

In general, as soon as a new problem arised, I added the mecanism to handle it, first locally, so, CLD first started with "P |= 0x08", which got turned into its own function when other flags had to be treated.

In retrospect, it was a very wise decision, and the architecture evolved with the issues. For instance, I moved from a giant switch to an array when there had too many entries, etc. I never designed any thing without beeing sure that I needed it. Even when I knew I needed something, I always waited for a real use case of the final form to implement it.

There have been two important decisions taken at the beginning, which had impact until the end. First I decided to count the cycles, so every instruction had cycle count included from the start. While there was no rationale for that apat from "it is easy, and it is cool", it proved to be absolutely fundamental at the end. The second decision was to include loggin in each instructions, so there was an automatic disassembly at execution. This decision was probably not the best, as it grouped the execution and display, which made sense when the only reason to ever display an instruction was to debug its execution. At a later stage, I could have used a disassembler on arbitrary data, but it was too difficult to separate it from execution.

D - The I/O memory

Pretty soon, it was visible that the ROM was trying to access the I/O memory zone. This zone is located around 0xC0nn (more or less), so I wrote specific code that stopped simulation if some unkown byte in that zone was touched. My worry was that the boot process would depend on the behavior of the perpheral attached to this zone.

Roughly, the code of the memory included:

bool is_io( const word adrs ) const { return adrs.h==0xC0; }

byte read_byte( const word adrs )
{
	if (is_io(adrs))
	{
		return io.read_byte(adrs.l);
	}
	return content[adrs];
}

io is a class that manages the I/O zone. In its original incarnation, it knew all the memory addresses name (as usual, only knowing the memory addresses that were used):

static std::string nameof( const byte adrs )
{
	switch (adrs)
	{
		case 0x51:
			return "TXTSET";
		default:
			std::cout << "??? Unknown adrs $C0" << to_hex2(adrs) << std::endl;
			assert( 0 );
	}
}

and 

byte read_byte( const byte adrs )
{
	std::cout << "IO READ @C0" << to_hex2(adrs) << "(" << nameof(adrs) << ")" << std::endl;

	if (adrs==0x51) { return 0x80; }

	return 0x00;
}

This simple design quickly got a few additional features crafted into it, and grew into a messy subsystem. This was somewhat expected, as the real needs here weren't clear at the beginning, and I didn't want to over-engineer. The ability to have I/O addresses displayed in symbolic form was a life saver, thought.

D - Addressing modes

Quickly, the same address modes used by multiple instructions started to pop up, so the need to share this code arised. 


a2_video.cpp
//
//  a2_video.cpp
//  FAE_SDL
//
//  Created by Frederic Stark on 30/05/16.
//  Copyright (c) 2016 Fred. All rights reserved.
//

#include "a2_video.hpp"
#include <cassert>

//	TEXT RENDERING

//	4 columns of 8x8 chars, columns first
//	@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_ !\"#$%&'()*+,-./0123456789:;<=>?
static const byte apple2font_bits[] = {
	0x00, 0x00, 0x00, 0x00, 0x1C, 0x1E, 0x00, 0x1C, 0x22, 0x22, 0x00, 0x22,
	0x2A, 0x22, 0x00, 0x32, 0x3A, 0x1E, 0x00, 0x2A, 0x1A, 0x02, 0x00, 0x26,
	0x02, 0x02, 0x00, 0x22, 0x3C, 0x02, 0x00, 0x1C, 0x00, 0x00, 0x00, 0x00,
	0x08, 0x1C, 0x08, 0x08, 0x14, 0x22, 0x08, 0x0C, 0x22, 0x22, 0x08, 0x08,
	0x22, 0x22, 0x08, 0x08, 0x3E, 0x2A, 0x08, 0x08, 0x22, 0x12, 0x00, 0x08,
	0x22, 0x2C, 0x08, 0x1C, 0x00, 0x00, 0x00, 0x00, 0x1E, 0x1E, 0x14, 0x1C,
	0x22, 0x22, 0x14, 0x22, 0x22, 0x22, 0x14, 0x20, 0x1E, 0x1E, 0x00, 0x18,
	0x22, 0x0A, 0x00, 0x04, 0x22, 0x12, 0x00, 0x02, 0x1E, 0x22, 0x00, 0x3E,
	0x00, 0x00, 0x00, 0x00, 0x1C, 0x1C, 0x14, 0x3E, 0x22, 0x22, 0x14, 0x20,
	0x02, 0x02, 0x3E, 0x10, 0x02, 0x1C, 0x14, 0x18, 0x02, 0x20, 0x3E, 0x20,
	0x22, 0x22, 0x14, 0x22, 0x1C, 0x1C, 0x14, 0x1C, 0x00, 0x00, 0x00, 0x00,
	0x1E, 0x3E, 0x08, 0x10, 0x22, 0x08, 0x3C, 0x18, 0x22, 0x08, 0x0A, 0x14,
	0x22, 0x08, 0x1C, 0x12, 0x22, 0x08, 0x28, 0x3E, 0x22, 0x08, 0x1E, 0x10,
	0x1E, 0x08, 0x08, 0x10, 0x00, 0x00, 0x00, 0x00, 0x3E, 0x22, 0x06, 0x3E,
	0x02, 0x22, 0x26, 0x02, 0x02, 0x22, 0x10, 0x1E, 0x1E, 0x22, 0x08, 0x20,
	0x02, 0x22, 0x04, 0x20, 0x02, 0x22, 0x32, 0x22, 0x3E, 0x1C, 0x30, 0x1C,
	0x00, 0x00, 0x00, 0x00, 0x3E, 0x22, 0x04, 0x38, 0x02, 0x22, 0x0A, 0x04,
	0x02, 0x22, 0x0A, 0x02, 0x1E, 0x22, 0x04, 0x1E, 0x02, 0x22, 0x2A, 0x22,
	0x02, 0x14, 0x12, 0x22, 0x02, 0x08, 0x2C, 0x1C, 0x00, 0x00, 0x00, 0x00,
	0x3C, 0x22, 0x08, 0x3E, 0x02, 0x22, 0x08, 0x20, 0x02, 0x22, 0x08, 0x10,
	0x02, 0x2A, 0x00, 0x08, 0x32, 0x2A, 0x00, 0x04, 0x22, 0x36, 0x00, 0x04,
	0x3C, 0x22, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x22, 0x22, 0x08, 0x1C,
	0x22, 0x22, 0x04, 0x22, 0x22, 0x14, 0x02, 0x22, 0x3E, 0x08, 0x02, 0x1C,
	0x22, 0x14, 0x02, 0x22, 0x22, 0x22, 0x04, 0x22, 0x22, 0x22, 0x08, 0x1C,
	0x00, 0x00, 0x00, 0x00, 0x1C, 0x22, 0x08, 0x1C, 0x08, 0x22, 0x10, 0x22,
	0x08, 0x14, 0x20, 0x22, 0x08, 0x08, 0x20, 0x3C, 0x08, 0x08, 0x20, 0x20,
	0x08, 0x08, 0x10, 0x10, 0x1C, 0x08, 0x08, 0x0E, 0x00, 0x00, 0x00, 0x00,
	0x20, 0x3E, 0x08, 0x00, 0x20, 0x20, 0x2A, 0x00, 0x20, 0x10, 0x1C, 0x08,
	0x20, 0x08, 0x08, 0x00, 0x20, 0x04, 0x1C, 0x08, 0x22, 0x02, 0x2A, 0x00,
	0x1C, 0x3E, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x22, 0x3E, 0x00, 0x00,
	0x12, 0x06, 0x08, 0x00, 0x0A, 0x06, 0x08, 0x08, 0x06, 0x06, 0x3E, 0x00,
	0x0A, 0x06, 0x08, 0x08, 0x12, 0x06, 0x08, 0x08, 0x22, 0x3E, 0x00, 0x04,
	0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x10, 0x02, 0x02, 0x00, 0x08,
	0x02, 0x04, 0x00, 0x04, 0x02, 0x08, 0x00, 0x02, 0x02, 0x10, 0x08, 0x04,
	0x02, 0x20, 0x08, 0x08, 0x3E, 0x00, 0x04, 0x10, 0x00, 0x00, 0x00, 0x00,
	0x22, 0x3E, 0x00, 0x00, 0x36, 0x30, 0x00, 0x00, 0x2A, 0x30, 0x00, 0x3E,
	0x2A, 0x30, 0x3E, 0x00, 0x22, 0x30, 0x00, 0x3E, 0x22, 0x30, 0x00, 0x00,
	0x22, 0x3E, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x22, 0x00, 0x00, 0x04,
	0x22, 0x00, 0x00, 0x08, 0x26, 0x08, 0x00, 0x10, 0x2A, 0x14, 0x00, 0x20,
	0x32, 0x22, 0x00, 0x10, 0x22, 0x00, 0x00, 0x08, 0x22, 0x00, 0x08, 0x04,
	0x00, 0x00, 0x00, 0x00, 0x1C, 0x00, 0x00, 0x1C, 0x22, 0x00, 0x20, 0x22,
	0x22, 0x00, 0x10, 0x10, 0x22, 0x00, 0x08, 0x08, 0x22, 0x00, 0x04, 0x08,
	0x22, 0x00, 0x02, 0x00, 0x1C, 0x7F, 0x00, 0x08,
};

typedef enum { kNormal,
	kInverted,
	kFlashing } eStyle;

//	Copies one scan line of text
static void text_scan_cpy(byte* const dest, const byte* const src, const int char_offset, const long frame)
{
	auto p      = src;
	uint32_t* q = (uint32_t*)dest;
	for (int x = 0; x != 40; x++)
	{
		char c = *p++;

		eStyle style;
		switch (c & 0xc0)
		{
			case 0x00:
				style = kInverted;
				break;
			case 0x40:
				style = kFlashing;
				break;
			default:
				style = kNormal;
				break;
		}

		if (style == kFlashing)
		{
			if (frame % 50 < 25)
			{
				style = kInverted;
			}
			else
			{
				style = kNormal;
			}
		}

		c &= 0x3f;

		byte b = apple2font_bits[((c & 0xf) << 5) + (c >> 4) + char_offset * 4];

		for (int x = 0; x != 7; x++)
		{
			*q++ = ((b & 0x1) ^ (style == kInverted)) ? 0xff40ff40 : 0xff000000;
			b >>= 1;
		}
	}
}

static uint32_t gr_color(const int color)
{
	auto c                      = color;
	static const int rgb[16][3] = {
		{ 0x00, 0x00, 0x00 }, //	BLACK
		{ 0xdd, 0x00, 0x33 }, //	"MAGENTA"
		{ 0x00, 0x00, 0x99 }, //	DARK BLUE
		{ 0xdd, 0x22, 0xdd }, //	PURPLE
		{ 0x00, 0x77, 0x22 }, //	DARK GREEN
		{ 0x55, 0x55, 0x55 }, //	GREY 1 (DARK)
		{ 0x22, 0x22, 0xff }, //	MEDIUM BLUE
		{ 0x66, 0xaa, 0xff }, //	LIGHT BLUE

		{ 0x88, 0x55, 0x00 }, //	BROWN
		{ 0xff, 0x66, 0x00 }, //	ORANGE
		{ 0xaa, 0xaa, 0xaa }, //	GREY 2 (LIGHT)
		{ 0xff, 0x99, 0x88 }, //	PINK
		{ 0x11, 0xdd, 0x00 }, //	LIGHT GREEN
		{ 0xff, 0xff, 0x00 }, //	YELLOW
		{ 0x44, 0xff, 0x99 }, //	AQUAMARINE
		{ 0xff, 0xff, 0xff }, //	WHITE
	};
	c &= 0xf;
	return (rgb[c][2] << 16) + (rgb[c][1] << 8) + rgb[c][0];
}

static void gr_scan_cpy(byte* const dest, const byte* const src, const bool high)
{
	uint32_t* q = (uint32_t*)dest;
	auto p      = src;
	for (int x = 0; x != 40; x++)
	{
		byte b = *p++;
		int c  = b >> (high * 4);
		for (int i = 0; i != 7; i++)
			*q++ = gr_color(c);
	}
}

static int text_offset(const int line)
{
	static const word offsets[24] = {
		0x000, 0x080, 0x100, 0x180, 0x200, 0x280, 0x300, 0x380,
		0x028, 0x0A8, 0x128, 0x1A8, 0x228, 0x2A8, 0x328, 0x3A8,
		0x050, 0x0D0, 0x150, 0x1D0, 0x250, 0x2D0, 0x350, 0x3D0
	};

	return offsets[line];
}

//	HIRES RENDERING

static int hires_offset(const int y)
{
	auto h = y;
	auto l = h & 0x7;
	h >>= 3;

	auto m = h & 0x7;
	h >>= 3;

	return h * 0x28 + m * 0x80 + l * 0x400;
}

static int mix(const int a, const int b)
{
	auto r = a + b;
	if (r > 255)
		r = 255;
	return r;
}

static void hires_scan_cpy(byte* const dst, const byte* const src)
{

	int last_color = 0;

	//	black, black, violet, green, black, black, blue, orange
	static const int red[8]   = { 0, 0, 255, 0, 0, 0, 0, 255 };
	static const int green[8] = { 0, 0, 0, 255, 0, 0, 128, 127 };
	static const int blue[8]  = { 0, 0, 255, 0, 0, 0, 255, 0 };

	auto q = dst;
	auto p = src;

	for (int x = 0; x != 40; x++)
	{
		auto b = *p++;

		const int type = (b & 0x80) >> 5; //	Type of byte (violet/green vs blue/orange)

		for (int i = 0; i != 7; i++)
		{
			const auto new_color = ((b & 0x1) << 1) | ((i & 0x1) ^ (x & 0x1)) | type;

			*q++ = mix(red[last_color], red[new_color]);
			*q++ = mix(green[last_color], green[new_color]);
			*q++ = mix(blue[last_color], blue[new_color]);
			*q++ = 0xff;

			last_color = new_color;
			b >>= 1;
		}
	}
}

void a2_video::draw(byte* const image, const long frame)
{
	const int base_hires = 0x2000 + is_page2() * 0x2000;
	const int base_text  = 0x0400 + is_page2() * 0x400;

	const byte* hires = memory_.dma_read(base_hires, 0x2000);
	const byte* text  = memory_.dma_read(base_text, 0x0400);

	assert(hires);
	assert(text);

	auto p = image;

	if (is_text())
		for (int y = 0; y != HGRMH; y++, p += HGRW * 4)
			text_scan_cpy(p, text + text_offset(y / 8), y & 0x7, frame);
	else if (is_hires())
		for (int y = 0; y != HGRMH; y++, p += HGRW * 4)
			hires_scan_cpy(p, hires + hires_offset(y));
	else
		for (int y = 0; y != HGRMH; y++, p += HGRW * 4)
			gr_scan_cpy(p, text + text_offset(y / 8), (y & 0x04) == 0x04);

	if (is_text() || is_mixed())
		for (int y = HGRMH; y != HGRH; y++, p += HGRW * 4)
			text_scan_cpy(p, text + text_offset(y / 8), y & 0x7, frame);
	else if (is_hires())
		for (int y = HGRMH; y != HGRH; y++, p += HGRW * 4)
			hires_scan_cpy(p, hires + hires_offset(y));
	else
		for (int y = HGRMH; y != HGRH; y++, p += HGRW * 4)
			gr_scan_cpy(p, text + text_offset(y / 8), (y & 0x04) == 0x04);
}
